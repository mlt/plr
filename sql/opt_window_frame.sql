create type tt as (x int, y int[]);
create or replace function fast_win_frame(r int, t tt) returns bool AS $$
  identical(parent.frame(), .GlobalEnv) && pg.throwerror('Parent env is global')
  exists('plr_window_frame', parent.frame(), inherits=FALSE) || pg.throwerror('No window frame data found')
  r == farg2[[prownum,2]][3]
$$ window language plr;
select s.r, s.p, fast_win_frame(NULLIF(r,4), (s.r, s.q)) over w
from (select r, r % 2 as p, array_fill(case when r=7 then 77 else r end, ARRAY[3]) as q from  generate_series(1,10) r) s
window w as (partition by p order by r rows between unbounded preceding and unbounded following)
order by s.r;