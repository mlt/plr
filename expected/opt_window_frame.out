create type tt as (x int, y int[]);
create or replace function fast_win_frame(r int, t tt) returns bool AS $$
  identical(parent.frame(), .GlobalEnv) && pg.throwerror('Parent env is global')
  exists('plr_window_frame', parent.frame(), inherits=FALSE) || pg.throwerror('No window frame data found')
  r == farg2[[prownum,2]][3]
$$ window language plr;
select s.r, s.p, fast_win_frame(NULLIF(r,4), (s.r, s.q)) over w
from (select r, r % 2 as p, array_fill(case when r=7 then 77 else r end, ARRAY[3]) as q from  generate_series(1,10) r) s
window w as (partition by p order by r rows between unbounded preceding and unbounded following)
order by s.r;
 r  | p | fast_win_frame
----+---+----------------
  1 | 1 | t
  2 | 0 | t
  3 | 1 | t
  4 | 0 |
  5 | 1 | t
  6 | 0 | t
  7 | 1 | f
  8 | 0 | t
  9 | 1 | t
 10 | 0 | t
(10 rows)

